index/ask/runner/runner.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return ask.Index{}, nil, err
+	}
+
+	rateLim := make(chan struct{}, qaRatelim)
+	var lock sync.Mutex
+	newAsks := make(map[string]ask.StorageAsk)
+	for i, addr := range addrs {
+		if ctx.Err() != nil {
+			break
+		}
+		rateLim <- struct{}{}
+		go func(addr address.Address) {
+			defer func() { <-rateLim }()
+			sask, ok, err := getMinerStorageAsk(ctx, api, addr)
+			if err != nil {
+				log.Errorf("getting miner storage ask: %s", err)
+				return
+			}
+			if !ok {
+				return
+			}
+			lock.Lock()
+			newAsks[addr.String()] = sask
+			lock.Unlock()
+		}(addr)
+		if i%100 == 0 {
+			stats.Record(context.Background(), metrics.MFullRefreshProgress.M(float64(i)/float64(len(addrs))))
+			log.Debugf("progress %d/%d", i, len(addrs))
+		}
+	}
+	for i := 0; i < qaRatelim; i++ {
+		rateLim <- struct{}{}
+	}
+
+	if ctx.Err() != nil {
+		return ask.Index{}, nil, fmt.Errorf("refresh was canceled")
+	}
+
+	stats.Record(context.Background(), metrics.MFullRefreshProgress.M(1))
+	ctx, _ = tag.New(context.Background(), tag.Insert(metrics.TagAskStatus, "FAIL"))
+	stats.Record(ctx, metrics.MAskQueryResult.M(int64(len(addrs)-len(newAsks))))
+	ctx, _ = tag.New(context.Background(), tag.Insert(metrics.TagAskStatus, "OK"))
+	stats.Record(ctx, metrics.MAskQueryResult.M(int64(len(newAsks))))
+
+	cache := generateOrderedAsks(newAsks)
+	return ask.Index{
+		LastUpdated:        time.Now(),
+		StorageMedianPrice: calculateMedian(cache),
+		Storage:            newAsks,
+	}, cache, nil
+}
+
+// getMinerStorage ask returns the result of querying the miner for its current Storage Ask.
+// If the miner has zero power, it won't be queried returning false.
+func getMinerStorageAsk(ctx context.Context, api *apistruct.FullNodeStruct, addr address.Address) (ask.StorageAsk, bool, error) {
+	ctx, cancel := context.WithTimeout(ctx, qaTimeout)
+	defer cancel()
+	power, err := api.StateMinerPower(ctx, addr, types.EmptyTSK)
+	if err != nil {
+		return ask.StorageAsk{}, false, fmt.Errorf("getting power %s: %s", addr, err)
+	}
+	if power.MinerPower.RawBytePower.IsZero() {
+		return ask.StorageAsk{}, false, nil
+	}
+	mi, err := api.StateMinerInfo(ctx, addr, types.EmptyTSK)
+	if err != nil {
+		return ask.StorageAsk{}, false, fmt.Errorf("getting miner %s info: %s", addr, err)
+	}
+
+	sask, err := api.ClientQueryAsk(ctx, mi.PeerId, addr)
+	if err != nil {
+		return ask.StorageAsk{}, false, nil
+	}
+	return ask.StorageAsk{
+		Miner:        sask.Ask.Miner.String(),
+		Price:        sask.Ask.Price.Uint64(),
+		MinPieceSize: uint64(sask.Ask.MinPieceSize),
+		Timestamp:    int64(sask.Ask.Timestamp),
+		Expiry:       int64(sask.Ask.Expiry),
+	}, true, nil
+}
+
+func calculateMedian(orderedAsks []*ask.StorageAsk) uint64 {
+	if len(orderedAsks) == 0 {
+		return 0
+	}
+	len := len(orderedAsks)
+	if len < 2 {
+		return orderedAsks[0].Price
+	}
+	if len%2 == 1 {
+		return orderedAsks[len/2].Price
+	}
+	return (orderedAsks[len/2-1].Price + orderedAsks[len/2].Price) / 2
+}
```

Comment:
This calculation takes advantage now of the orderedAsk that is maintained of Storage Ask by price.
Before it was sorting and doing the calculation. Nothing fancy... just leveraging that we had that data sorted for other purpose.